### PR TITLE
Cache system version 2

### PR DESCRIPTION
**Pull request by meooow25**
### Changes
- A new CacheSystem with the components:
  - ContestCache maintains contests. Updates at regular intervals. Update delay is decreased when a contest is active.
  - ProblemCache maintains problems. Updates at regular intervals.
  - RatingChangesCache is responsible for fetching rating changes of any contest that just had rating changes applied.
  - RanklistCache is responsible for fetching ranklists of any active contests.
- A simple EventSystem.
- A CacheControl cog that allows manual update of cache.
- A new table in the database and new methods in HandleConn to store/retreive rating changes to/from the database.

### CODE REVIEW

File: tle/util/cache_system2.py [hunk 1]
```diff
@@ -0,0 +1,477 @@
...
+    def __init__(self, contest, problems, standings, get_current_rating, fetch_time):
+        self.contest = contest
+        self.problems = problems
+        self.standings = standings
+        self.fetch_time = fetch_time
+        self.delta_by_handle = {}
+        self.prepare_predictions()
+
+    def prepare_predictions(self):
+        # TODO: Implement
```
Comment (algmyr) [inline tle/util/cache_system2.py]: Not for this PR, I presume?

File: tle/util/cache_system2.py [hunk 2]
```diff
@@ -0,0 +1,477 @@
...
+                await cf_common.event_sys.wait_for('EVENT_CONTEST_LIST_REFRESH')
+                await self._process_contests()
+            except Exception:
+                self.logger.warning('Exception in ranklist updater task, ignoring.', exc_info=True)
+
+    async def _process_contests(self):
+        contests_by_phase = self.cache_master.contest_cache.contests_by_phase
+        running_contests = contests_by_phase['_RUNNING']
+        check = self.cache_master.rating_changes_cache.is_newly_finished_without_rating_changes
+        to_monitor = running_contests + [contest for contest in contests_by_phase['FINISHED'] if check(contest)]
```
Comment (algmyr) [inline tle/util/cache_system2.py]: consider `list(filter(...))`

Comment (meooow25) [reply tle/util/cache_system2.py]: Done

File: tle/util/cache_system2.py [hunk 3]
```diff
@@ -0,0 +1,477 @@
...
+        new_ids = {contest.id for contest in to_monitor}
+        if new_ids != self.ranklist_by_contest.keys():
+            if self.update_task is not None:
```
[the remark this answers is not in the document]
Comment (meooow25) [reply tle/util/cache_system2.py]: Done

File: tle/util/cache_system2.py [hunk 4]
```diff
@@ -0,0 +1,477 @@
...
+        await self._update(problems)
+
+    async def _update(self, problems):
+        self.logger.info(f'{len(problems)} problems fetched from API')
+        banned_tags = ['*special']
+        contest_map = {problem.contestId: self.cache_master.contest_cache.contest_by_id.get(problem.contestId)
+                       for problem in problems}
+
+        def keep(problem):
+            return contest_map[problem.contestId] is not None and problem.has_metadata() \
```
[the remark this answers is not in the document]
Comment (meooow25) [reply tle/util/cache_system2.py]: Done

File: tle/util/cache_system2.py [hunk 5]
```diff
@@ -0,0 +1,477 @@
...
+        reloading = self.reload_lock.locked()
+        if reloading:
+            # Wait until reload complete.
+            # To wait until lock is free, await acquire then release immediately.
+            async with self.reload_lock:
+                pass
+        else:
+            await self._pre_reload()
+
+        if self.reload_exception is not None:
```
[the remark this answers is not in the document]
Comment (meooow25) [reply tle/util/cache_system2.py]: Done

File: tle/util/cache_system2.py [hunk 6]
```diff
@@ -0,0 +1,477 @@
...
```
[the remark this answers is not in the document]
Comment (meooow25) [reply tle/util/cache_system2.py]: Done

File: tle/util/cache_system2.py [hunk 7]
```diff
@@ -0,0 +1,477 @@
...
+        # between the two FINISHED phases, we are forced to fetch during both.
+        # A contest also has empty list if it is unrated. We assume that is the case if
+        # _RATED_DELAY time has passed since the contest end.
+
+        to_monitor = [contest for contest in self.cache_master.contest_cache.contests_by_phase['FINISHED']
+                      if self.is_newly_finished_without_rating_changes(contest)]
+        cur_ids = {contest.id for contest in self.monitored_contests}
+        new_ids = {contest.id for contest in to_monitor}
+        if new_ids != cur_ids:
+            if self.update_task is not None:
```
[the remark this answers is not in the document]
Comment (meooow25) [reply tle/util/cache_system2.py]: Done

File: tle/util/cache_system2.py [hunk 8]
```diff
@@ -0,0 +1,477 @@
...
+        self.logger.info(f'{len(problems)} problems fetched from API')
+        banned_tags = ['*special']
+        contest_map = {problem.contestId: self.cache_master.contest_cache.contest_by_id.get(problem.contestId)
+                       for problem in problems}
+
+        def keep(problem):
+            return contest_map[problem.contestId] is not None and problem.has_metadata() \
+                   and not problem.tag_matches(banned_tags)
+
+        filtered_problems = [problem for problem in problems if keep(problem)]
```
[the remark this answers is not in the document]
Comment (meooow25) [reply tle/util/cache_system2.py]: Done

File: tle/util/cache_system2.py (hunk 1 above)
Comment (meooow25) [reply tle/util/cache_system2.py]: Nope, this will be a whole new feature!

**Pull request merged**